src/globus_sdk/experimental/globus_app/app.py

```diff
@@ -224,8 +224,67 @@ def _initialize_authorizer_factory(self) -> None:
         authorize requests.
         """
 
+    def login(
+        self,
+        *,
+        auth_params: GlobusAuthorizationParameters | None = None,
+        force: bool = False,
+    ) -> None:
+        """
+        Log an auth entity into the app, if needed, storing the resulting tokens.
+
+        A login flow will be is run if:
+            * `auth_params` are provided
+            * `force` is set to True
+            * `self.login_required()` evaluates to True
+
+        :param auth_params: An optional set of authorization parameters to establish
+            requirements and controls for the login flow.
+        :param force: If True, run a login flow even if one is not necessary.
+        """
+        if auth_params or force or self.login_required():
+            self._run_login_flow(auth_params)
+
+    def login_required(self) -> bool:
+        """
+        Determine if a login flow is required to interact with resource servers under
+        the current scope requirements.
+
+        This will return false if:
+            * access tokens have never been issued
+            * access tokens have been issued but are underscoped
+            * access tokens have expired and wouldn't be resolved with refresh tokens
+
+        :returns: True if a login flow is required, False otherwise.
+        """
+        for resource_server in self._scope_requirements.keys():
+            try:
+                self.get_authorizer(resource_server, skip_error_handling=True)
+            except TokenValidationError:
+                return True
+        return False
+
+    def logout(self) -> None:
+        """
+        Log an auth entity out of the app.
+
+        This will remove and revoke all tokens stored for the current app user.
+        """
+        # Revoke all tokens, removing them from the underlying token storage
+        for resource_server in self._scope_requirements.keys():
+            token_storage = self._validating_token_storage.token_storage
+            token_data = token_storage.get_token_data(resource_server)
+            if token_data:
+                self._login_client.oauth2_revoke_token(token_data.access_token)
+                if token_data.refresh_token:
+                    self._login_client.oauth2_revoke_token(token_data.refresh_token)
+                token_storage.remove_token_data(resource_server)
+
+        # Invalidate any cached authorizers
+        self._authorizer_factory.clear_cache()
+
     @abc.abstractmethod
-    def run_login_flow(
+    def _run_login_flow(
```

Comment:
This docstring is incomplete. I can't comment directly on the affected line to provide a suggestion

Reply:
Copied the text over from `login`'s docstring.